### PR TITLE
fix(interface): identity + data subject

### PR DESCRIPTION
**Pull request by smalexan**
## Description of this change

Add additional identity metadata
DSR data subject info
Handle data subject user input request
Consent priority

## Why is this change being made?
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have evaluated the security impact of this change, and [OWASP Secure Coding Practices](https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/migrated_content#) have been observed.
- [x] I have informed stakeholders of my changes.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -156,6 +161,12 @@ export interface InvokeRightRequest {
   jurisdictionCode: string
   rightCode: string
   user: User
+
+  // dataSubjectTypeCode is the identifier representing the data subject type specified by the user
+  dataSubjectTypeCode: string
+
+  // dataSubjectTypeUserInput is additional information provided by the user describing their relation to the business
+  dataSubjectTypeUserInput: string
```
Comment (sethyates) [inline src/index.ts:169]: Why is this not in `User` (which we should rename to `DataSubject`)?

Comment (smalexan) [reply src/index.ts:169]: ah oops it should be

Comment (smalexan) [reply src/index.ts]: fixed

File: src/index.ts [hunk 2]
```diff
@@ -144,6 +149,12 @@ export interface User {
   postalCode?: string
   addressLine1?: string
   addressLine2?: string
+
+  // dataSubjectTypeCode is the identifier representing the data subject type specified by the user
+  dataSubjectTypeCode?: string
```
Comment (sethyates) [inline src/index.ts:154]: `typeCode`?

File: src/index.ts [hunk 3]
```diff
@@ -144,6 +149,12 @@ export interface User {
...
+
+  // dataSubjectTypeUserInput is additional information provided by the user describing their relation to the business
+  dataSubjectTypeUserInput?: string
```
Comment (sethyates) [inline src/index.ts:157]: not sure what this means.

File: src/index.ts [hunk 4]
```diff
@@ -248,9 +259,49 @@ export interface CanonicalPurpose {
...
+// IdentityLocation is the location on the page from which to retrieve identity information
+export enum IdentityLocation {
+  IDENTITY_LOCATION_UNDEFINED,
+  IDENTITY_LOCATION_DATA_LAYER,
+  IDENTITY_LOCATION_WINDOW,
+  IDENTITY_LOCATION_COOKIE,
+  IDENTITY_LOCATION_MANAGED, // this is created if necessary and stored in a cookie with the associated name
+  IDENTITY_LOCATION_LOCAL_STORAGE,
+  IDENTITY_LOCATION_SESSION_STORAGE,
+  IDENTITY_LOCATION_URL_QUERY_PARAM,
```
Comment (sethyates) [inline src/index.ts:271]: QUERY_STRING

File: src/index.ts (hunk 3 above)
Comment (smalexan) [reply src/index.ts:157]: We want to support data subject types that require more information from the user. 

A use case would be an "Other" data subject type where below that there is a text box for a user to add more information about their relation to the business. This field would send what the user enters in this text box

File: src/index.ts [hunk 5]
```diff
@@ -144,6 +149,12 @@ export interface User {
   postalCode?: string
   addressLine1?: string
   addressLine2?: string
+
+  // typeCode is the identifier representing the data subject type specified by the user
+  typeCode?: string
+
+  // typeDescription is additional information provided by the user describing their relation to the business
+  typeDescription?: string
```
Comment (sethyates) [inline src/index.ts:157]: how is this different to the `description` above?

File: src/index.ts [hunk 6]
```diff
@@ -248,9 +259,49 @@ export interface CanonicalPurpose {
...
+  IDENTITY_FORMAT_SEMICOLON,
+}
+
+// Identity represents all the metadata for an identifier on the page
 export interface Identity {
-  type: string
-  variable: string
+
+  // DEPRECATED: type is a string representing the location of the identity on the page
+  type?: string // deprecated
```
[the remark this answers is not in the document]
Comment (sethyates) [reply src/index.ts:288]: asking, not advocating

File: src/index.ts [hunk 7]
```diff
@@ -248,9 +259,49 @@ export interface CanonicalPurpose {
...
+
+  // location on the page from which to retrieve identity information
+  location?: IdentityLocation
```
Comment (sethyates) [inline src/index.ts:291]: what is the default then if null? should this be required?

File: src/index.ts (hunk 5 above)
Comment (smalexan) [reply src/index.ts:157]: It is a separate text box directly under the data subject dropdown that specifically is for information about how the data subject related to the business. Don't think there are screens here yet but asked in figma

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (smalexan) [reply src/index.ts:288]: yeah can keep as type and use string enums

File: src/index.ts (hunk 7 above)
Comment (smalexan) [reply src/index.ts:291]: yeah should be required - just not there yet. will keep type

File: src/index.ts [hunk 8]
```diff
@@ -248,9 +259,46 @@ export interface CanonicalPurpose {
...
+  IDENTITY_FORMAT_SEMICOLON = "semicolon",
+}
+
+// Identity represents all the metadata for an identifier on the page
 export interface Identity {
-  type: string
-  variable: string
+
+  // type is the location on the page from which to retrieve identity information
+  type: IdentityType
```
Comment (smalexan) [inline src/index.ts:288]: @sethyates think the string equality checking we do now will be compatible with this

**Pull request merged**